packages/http-server/src/getHttpConfigFromRequest.ts

```diff
@@ -1,26 +1,37 @@
-import { IHttpOperationConfig, IHttpRequest } from '@stoplight/prism-http';
+import { IHttpOperationConfig, IHttpRequest, ProblemJsonError, UNPROCESSABLE_ENTITY } from '@stoplight/prism-http';
+import { pipe } from 'fp-ts/lib/pipeable';
+import * as E from 'fp-ts/lib/Either';
+import * as t from 'io-ts';
+import { failure } from 'io-ts/lib/PathReporter';
+import { BooleanFromString } from 'io-ts-types/lib/BooleanFromString';
+//@ts-ignore
+import * as parsePreferHeader from 'parse-prefer-header';
 
-export const getHttpConfigFromRequest = (req: IHttpRequest): Partial<IHttpOperationConfig> => {
-  const httpOperationConfig: Partial<IHttpOperationConfig> = {};
-  const query = req.url.query;
+const PreferencesDecoder = t.union([
+  t.undefined,
+  t.partial(
+    {
+      code: t.string,
+      dynamic: t.string.pipe(BooleanFromString),
+      example: t.string,
+    },
+    'Preferences'
+  ),
+]);
 
-  if (!query) {
-    return {};
-  }
+type requestPreference = Partial<Omit<IHttpOperationConfig, 'mediaType'>>;
 
-  const { __code, __dynamic, __example } = query;
+export const getHttpConfigFromRequest = (req: IHttpRequest): E.Either<Error, requestPreference> => {
+  const preferences =
+    req.headers && req.headers['prefer']
```

Comment:
Thinking about two things here:
1. if `Prefer` header exists then all values passed via query are ignored. So:
`
GET /?__dynamic=false
Prefer: code=202
`
Will result in `dynamic=true; code=202`.

So, there are few questions. Do we want to force-move people to Prefer header? If so, then leaving it as is have educational value. If not, then we should support mixture of query and header with header preference.


2. I believe we should be a bit more careful about this header. Our power-api-expert-enterprise users may already use Prefer header and have it defined in their API descriptions (e.g. with regex pattern). Our headers can interfere with theirs leading to unexpected 422s.

I would extend the approach with stripping "our" preferences from header and if it is empty after that, removing it at all. That should happen before further processing in Prism.

Reply:
1. This is correct — if somebody specifies the options in the header, they will take the full preference and and the query is going to be ignored. I'll update the documentation so that this stands clear.

2. That is really really unlikely to be happening. The `Prefer` header is a multi key value and multi purpose header and the changes that some other proxies (Cloudflare, Netlify, Fastly or whatever is fancy these days) are using/modifying it is a real possibility; this means that this is one of the few headers that you can't really validate for real if you do not parse it first — thing that almost nobody is doing. My take is that for now we keep it as it is (no strip, no parse) and if somebody will start to complain, we will figure out a solution for it.

Reply:
> This is correct — if somebody specifies the options in the header, they will take the full preference and and the query is going to be ignored. I'll update the documentation so that this stands clear.

It is minor, but if you find a suitable place for this information it would be good to add it.

Reply:
> That is really really unlikely to be happening. The Prefer header is a multi key value and multi purpose header and the changes that some other proxies (Cloudflare, Netlify, Fastly or whatever is fancy these days) are using/modifying it is a real possibility; this means that this is one of the few headers that you can't really validate for real if you do not parse it first — thing that almost nobody is doing. My take is that for now we keep it as it is (no strip, no parse) and if somebody will start to complain, we will figure out a solution for it.

I agree. We are not closing the doors for fixing it in the future.